### PR TITLE
feat(ext/crypto): support importing ECSDA and ECDH

### PR DESCRIPTION
**Pull request by SeanWykes**
importKey implemented for ECDSA and ECDH, extracted from #13013 with modifications
- pkcs8 : curve P-256 only
- spki : curves P-256 and P-384
- jwk : 
    public key - curves P-256 and P-384
    private key - curve P-256 only

Restrictions on to P-384 curve due to lack of support in p384 crate.

### CODE REVIEW

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2501]: This is already validated 30 lines above, and can thus be removed here.

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2231]: Ordering is important, because this is observable. This check happens as the first step for the "raw" algorithm, but only later for JWK.

Please move this to all the relevant spots in the different algorithms.

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2349]:
```suggestion
        const keyType = (jwk.d !== undefined) ? "private" : "public";
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2364]:
```suggestion
            "'kty' property of JsonWebKey must be 'EC'",
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2370]:
```suggestion
        if (keyUsages.length > 0 && jwk.use !== undefined) {
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2371]: You can collapse these two if statements into one.

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2374]:
```suggestion
              `'use' property of JsonWebKey must be '${supportedUsages.jwkUse}'`,
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2382]:
```suggestion
        if (jwk.key_ops !== undefined) {
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2409]:
```suggestion
        if (jwk.ext === false && extractable === true) {
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2411]:
```suggestion
            "'ext' property of JsonWebKey must not be false if extractable is true",
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2417]:
```suggestion
        if (jwk.alg !== undefined && normalizedAlgorithm.name == "ECDSA") {
```

File: ext/crypto/00_crypto.js
Comment (lucacasonato) [inline ext/crypto/00_crypto.js:2443]:
```suggestion
                "Mismatched curve algorithm",
```

File: ext/crypto/01_webidl.js
Comment (lucacasonato) [inline ext/crypto/01_webidl.js:133]:
```suggestion
    .createDictionaryConverter("EcKeyGenParams", dictEcKeyGenParams);

  const dictEcImportParams = [
    ...dictAlgorithm,
```

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:15]:
```suggestion
use pkcs8::FromPrivateKey;
use pkcs8::ToPrivateKey;
```

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:19]:
```suggestion
use p256::elliptic_curve::generic_array::typenum::U32;
use p256::elliptic_curve::generic_array::ArrayLength;
use p256::elliptic_curve::generic_array::GenericArray;
```

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:738]: p384 does implement this. `p384::SecretKey` is a `elliptic_curve::SecretKey`, which has a `to_pkcs8_der` method: https://docs.rs/elliptic-curve/0.11.5/elliptic_curve/struct.SecretKey.html#method.to_pkcs8_der.

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:679]:
```suggestion
  jwt_b64_int_or_err!(val, b64, "invalid b64 coordinate");
```

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:758]: This is specific to p256.

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:814]: This needs to be done before landing.

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:828]: I don't think this PKCS8 importing is quite correct. This should look more like the RSA-PSS PKCS8 importing. I.e. first parse to a `pkcs8::PrivateKeyInfo`, then validate the OID + parameters as per spec etc.

The check in the spec that confirms that the named curve the user specified, and the one in the DER match is not present for example.

File: ext/crypto/import_key.rs
Comment (lucacasonato) [inline ext/crypto/import_key.rs:836]: Same goes for SPKI. There should be a check here that checks the algorithm OID and the namedCurve are correct. Look at how this is done for RSA-PSS (checking OID and parameters).

**Review by lucacasonato — COMMENTED** [review]
Good start. Thanks @SeanWykes.

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs]: @lucacasonato The p384 is missing relevant functionality, specifically the ProjectiveArithmetic trait, which is a dependency for to_pkcs8_der().

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs]: Missing ProjectiveArithmetic in p384 crate. For now, have changed code to throw an "Unsupported curve" error, instead of blindly importing the PKCS8.

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs]: OK. Thanks for the hint. I've been focussing on functionality first .. now getting into the validation stuff. Just committed for SPKI, PKCS8 next on list.

File: ext/crypto/import_key.rs
Comment (lucacasonato) [reply ext/crypto/import_key.rs]: :+1:

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs:907]: done. A bit messy 'cos `spki` and `pkcs8` crates use different versions of `der::asn1` constants, so transformed my OID constants into strings and doing string compares which I know is not optimal. Any ideas?

One deviation from spec, in step 10 is that I do not explicitly parse the pkcs#1 private_key DER from the PKCS8, just pass the pkcs8 blob strait to `SecretKey` during checking step 12, assuming that SecretKey with validate the pkcs1 structure as it extracts `d`.

Finally, as for step 12 (validating that the key is a valid point on the curve, any EC experts around to chip in? @tarcieri?

File: ext/crypto/Cargo.toml
Comment (lucacasonato) [inline ext/crypto/Cargo.toml:25]: What do you need the `pem` feature for?

File: ext/crypto/Cargo.toml
Comment (lucacasonato) [inline ext/crypto/Cargo.toml:27]: What do you need this for? It should be re-exported from various crates.

**Review by lucacasonato — COMMENTED** [review]
Are the version bumps necessary? I would rather wait for them until `rsa` has gotten a new release, so we don't duplicate so many dependencies.

**Comment by SeanWykes** [comment]
> Are the version bumps necessary? I would rather wait for them until `rsa` has gotten a new release, so we don't duplicate so many dependencies.

Not as such .. bumped during attempts to improve p384 coverage (pkcs8 etc). Will back-out the changes ..

**Comment by lucacasonato** [comment]
@SeanWykes I have unbumped the crates locally - will push that shortly.

File: ext/crypto/Cargo.toml
Comment (SeanWykes) [reply ext/crypto/Cargo.toml:25]: Don't need PEM. It was the only way I could find to satisfy constraint for trait `FromPrivateKey` to use `from_pkcs8_der` on `p256::SecretKey<NistP256>`. Line ~730 import_key.rs

Comment (lucacasonato) [reply ext/crypto/Cargo.toml]: Yeah, have unfortunately discovered this myself now during testing.

**Review by lucacasonato — COMMENTED** [review]
There are some WPT regressions it seems that I need to be dealt with before merge.

I will try to get this landed for 1.17.0 tomorrow. If it doesn't make it, I'll try to land it for 1.17.1 in a week from now. Maybe we can have web crypto "done" by the end of the year.

**Comment by tarcieri** [comment]
I can take a look at getting the `rsa` crate bumped to the latest versions of `pkcs1`/`pkcs8`

File: ext/crypto/import_key.rs
Comment (tarcieri) [reply ext/crypto/import_key.rs:905]: > A bit messy 'cos spki and pkcs8 crates use different versions of der::asn1 constants, so transformed my OID constants into strings and doing string compares which I know is not optimal.

Can you be more specific about the problems you’re having here? Code examples and compiler errors would be helpful

**Comment by SeanWykes** [comment]
> There are some WPT regressions it seems that I need to be dealt with before merge. (I can do this)
> 
> I will try to get this landed for 1.17.0 tomorrow. If it doesn't make it, I'll try to land it for 1.17.1 in a week from now. Maybe we can have web crypto "done" by the end of the year.

@lucacasonato WPT - import_export/ec will only pass the supported cases when #13104 is in, since the tests need exportKey to function. PTAL @ #13104 it should hopefully be in reasonable shape. ;)

**Comment by lucacasonato** [comment]
> import_export/ec will only pass the supported cases when #13104 is in

I am aware. I meant the cases in `WebCryptoAPI/sign_verify` (I have no idea yet what is going on there).

**Comment by SeanWykes** [comment]
> > import_export/ec will only pass the supported cases when #13104 is in
> 
> I am aware. I meant the cases in `WebCryptoAPI/sign_verify` (I have no idea yet what is going on there).
@lucacasonato, I think there maybe something strange going on with key-hash combinations. In `op_crypto_sign_key ECDSA`, the comment states underlying `ring::signature` crate only supports two algo pairs 
      // We only support P256-SHA256 & P384-SHA384. These are recommended signature pairs.
      // https://briansmith.org/rustdoc/ring/signature/index.html#statics
However, the hash param passed from js is checked (SHA-256 or SHA-384), independent of the curve used.

**Comment by lucacasonato** [comment]
Yeah, just found that too. This code needs a general rewrite in the future.

I'll clean up this PR now, and then land it.

**Review by lucacasonato — APPROVED** [review]
LGTM, thanks @SeanWykes.

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs:757]: yes. can clean-up when p384 support implemented, maybe using generic impl. as per work in progress on Rust Crypto [here](https://github.com/RustCrypto/elliptic-curves/issues/240)

File: ext/crypto/import_key.rs
Comment (SeanWykes) [inline ext/crypto/import_key.rs:859]: // id-secp521r1

File: ext/crypto/import_key.rs
Comment (SeanWykes) [reply ext/crypto/import_key.rs:904]: @tarcieri - code fixed by @lucacasonato, so no longer relevant to PR. 
I was getting compiler error "maybe different versions of `const_oids` due to trying to mix `ObjectIdentifier` from `pkcs8::der' and `spki::der'.
Thanks.

**Comment by SeanWykes** [comment]
> LGTM, thanks @SeanWykes.

@lucacasonato - thanks for your help, cleanups and fixes. Learnt a lot about paradigmatic Rust from this exercise.

**Pull request merged**
